File: app/src/main/java/com/majeur/psclient/model/StatModifiers.java
package com.majeur.psclient.model;

import static com.majeur.psclient.util.Utils.boldText;
import static com.majeur.psclient.util.Utils.str;

public class StatModifiers {

    public static final String[] STAT_KEYS = {"atk", "def", "spa", "spd", "spe", "evasion", "accuracy"};

    private static final float[] LEVELS = {1f / 4f, 2f / 7f, 1f / 3f, 2f / 5f, 1f / 2f, 2f / 3f, 1f, 3f / 2f, 2f, 5f / 2f, 3f, 7f / 2f, 4f};
    private static final float[] LEVELS_ALT = {3f / 9f, 3f / 8f, 3f / 7f, 3f / 6f, 3f / 5f, 3f / 4f, 3f / 3f, 4f / 3f, 5f / 3f, 6f / 3f, 7f / 3f, 8f / 3f, 9f / 3f};

    private int atk = 0;
    private int def = 0;
    private int spa = 0;
    private int spd = 0;
    private int spe = 0;
    private int eva = 0;
    private int acc = 0;

    public int get(String stat) {
        switch (stat) {
            case "atk":
                return atk;
            case "def":
                return def;
            case "spa":
                return spa;
            case "spd":
                return spd;
            case "spe":
                return spe;
            case "evasion":
                return eva;
            case "accuracy":
                return acc;
            default:
                return 0;
        }
    }

    public void inc(String stat, int val) {
        switch (stat) {
            case "atk":
                atk += val;
                break;
            case "def":
                def += val;
                break;
            case "spa":
                spa += val;
                break;
            case "spd":
                spd += val;
                break;
            case "spe":
                spe += val;
                break;
            case "evasion":
                eva += val;
                break;
            case "accuracy":
                acc += val;
                break;
        }
    }

    public void set(String stat, int val) {
        switch (stat) {
            case "atk":
                atk = val;
                break;
            case "def":
                def = val;
                break;
            case "spa":
                spa = val;
                break;
            case "spd":
                spd = val;
                break;
            case "spe":
                spe = val;
                break;
            case "evasion":
                eva = val;
                break;
            case "accuracy":
                acc = val;
                break;
        }
    }

    public void set(StatModifiers modifiers) {
        atk = modifiers.atk;
        def = modifiers.def;
        spa = modifiers.spa;
        spd = modifiers.spd;
        spe = modifiers.spe;
        eva = modifiers.eva;
        acc = modifiers.acc;
    }

    public void invert() {
        atk = -atk;
        def = -def;
        spa = -spa;
        spd = -spd;
        spe = -spe;
        eva = -eva;
        acc = -acc;
    }

    public void clear() {
        atk = def = spa = spd = spe = eva = acc = 0;
    }

    public void clearPositive() {
        if (atk > 0) atk = 0;
        if (def > 0) def = 0;
        if (spa > 0) spa = 0;
        if (spd > 0) spd = 0;
        if (spe > 0) spe = 0;
        if (eva > 0) eva = 0;
        if (acc > 0) acc = 0;
    }

    public void clearNegative() {
        if (atk < 0) atk = 0;
        if (def < 0) def = 0;
        if (spa < 0) spa = 0;
        if (spd < 0) spd = 0;
        if (spe < 0) spe = 0;
        if (eva < 0) eva = 0;
        if (acc < 0) acc = 0;
    }

    public float modifier(String stat) {
        switch (stat) {
            case "atk":
                return LEVELS[atk + 6];
            case "def":
                return LEVELS[def + 6];
            case "spa":
                return LEVELS[spa + 6];
            case "spd":
                return LEVELS[spd + 6];
            case "spe":
                return LEVELS[spe + 6];
            case "evasion":
                return LEVELS_ALT[eva + 6];
            case "accuracy":
                return LEVELS_ALT[acc + 6];
            default:
                return 0;
        }
    }

    public CharSequence calcReadableStat(String stat, int baseStat) {
        float m = modifier(stat);
        if (m == 1f) return str(baseStat);
        int afterModifier = (int) (baseStat * m);
        return boldText(str(afterModifier));
    }
}

File: psclient/src/main/java/com/majeur/psclient/model/battle/StatModifiers.kt
package com.majeur.psclient.model.battle

import com.majeur.psclient.util.bold

class StatModifiers {

    private var atk = 0
        set(value) { field = value.coerceIn(-6, 6) }
    private var def = 0
        set(value) { field = value.coerceIn(-6, 6) }
    private var spa = 0
        set(value) { field = value.coerceIn(-6, 6) }
    private var spd = 0
        set(value) { field = value.coerceIn(-6, 6) }
    private var spe = 0
        set(value) { field = value.coerceIn(-6, 6) }
    private var eva = 0
        set(value) { field = value.coerceIn(-6, 6) }
    private var acc = 0
        set(value) { field = value.coerceIn(-6, 6) }

    operator fun get(stat: String?) = when (stat) {
        "atk" -> atk
        "def" -> def
        "spa" -> spa
        "spd" -> spd
        "spe" -> spe
        "evasion" -> eva
        "accuracy" -> acc
        else -> 0
    }

    fun inc(stat: String?, value: Int) {
        when (stat) {
            "atk" -> atk += value
            "def" -> def += value
            "spa" -> spa += value
            "spd" -> spd += value
            "spe" -> spe += value
            "evasion" -> eva += value
            "accuracy" -> acc += value
        }
    }

    operator fun set(stat: String?, value: Int) {
        when (stat) {
            "atk" -> atk = value
            "def" -> def = value
            "spa" -> spa = value
            "spd" -> spd = value
            "spe" -> spe = value
            "evasion" -> eva = value
            "accuracy" -> acc = value
        }
    }

    fun set(modifiers: StatModifiers) {
        atk = modifiers.atk
        def = modifiers.def
        spa = modifiers.spa
        spd = modifiers.spd
        spe = modifiers.spe
        eva = modifiers.eva
        acc = modifiers.acc
    }

    fun invert() {
        atk = -atk
        def = -def
        spa = -spa
        spd = -spd
        spe = -spe
        eva = -eva
        acc = -acc
    }

    fun clear() {
        acc = 0
        eva = 0
        spe = 0
        spd = 0
        spa = 0
        def = 0
        atk = 0
    }

    fun clearPositive() {
        if (atk > 0) atk = 0
        if (def > 0) def = 0
        if (spa > 0) spa = 0
        if (spd > 0) spd = 0
        if (spe > 0) spe = 0
        if (eva > 0) eva = 0
        if (acc > 0) acc = 0
    }

    fun clearNegative() {
        if (atk < 0) atk = 0
        if (def < 0) def = 0
        if (spa < 0) spa = 0
        if (spd < 0) spd = 0
        if (spe < 0) spe = 0
        if (eva < 0) eva = 0
        if (acc < 0) acc = 0
    }

    fun modifier(stat: String?): Float = when (stat) {
        "atk" -> LEVELS[atk + 6]
        "def" -> LEVELS[def + 6]
        "spa" -> LEVELS[spa + 6]
        "spd" -> LEVELS[spd + 6]
        "spe" -> LEVELS[spe + 6]
        "evasion" -> LEVELS_ALT[eva + 6]
        "accuracy" -> LEVELS_ALT[acc + 6]
        else -> 0f
    }


    fun calcReadableStat(stat: String?, baseStat: Int): CharSequence {
        val m = modifier(stat)
        if (m == 1f) return baseStat.toString()
        val afterModifier = (baseStat * m).toInt()
        return afterModifier.toString().bold()
    }

    companion object {
        @JvmStatic val STAT_KEYS = arrayOf("atk", "def", "spa", "spd", "spe", "evasion", "accuracy")
        private val LEVELS = floatArrayOf(1f / 4f, 2f / 7f, 1f / 3f, 2f / 5f, 1f / 2f, 2f / 3f, 1f, 3f / 2f, 2f, 5f / 2f, 3f, 7f / 2f, 4f)
        private val LEVELS_ALT = floatArrayOf(3f / 9f, 3f / 8f, 3f / 7f, 3f / 6f, 3f / 5f, 3f / 4f, 3f / 3f, 4f / 3f, 5f / 3f, 6f / 3f, 7f / 3f, 8f / 3f, 9f / 3f)
    }
}
